TokenFetcher support to get authorizationHeader - RC.3.0

aboktor opened the pull request:
## Description
There is an assumption throughout the odsp-driver code that tokens are always Bearer tokens. This is not the case. AT_POP tokens could also be used. AT_POP tokens encode url path, query params and http method into the token. This means that the token itself cannot be passed via query params (otherwise that would be a circular dependency).

This PR adds authorizationHeader to TokenResponse as well as accepts request on the TokenFetcher interface.

## Breaking Changes
None, all changes to public interfaces should be backwards compatible.
Behavior changes:
- forceAccessTokenViaAuthorizationHeader is now a no-op.
- ICollabSessionOptions.unauthenticatedUserDisplayName is no longer used or passed through joinSession

## Reviewer Guidance
Nothing atm

aboktor committed:
Adding request to TokenFetchOptions and authorizationHeader to TokenResponse
Made all necessary changes to consume the new interface

aboktor committed:
Not sure why these changed

AndreiZe commented on packages/drivers/odsp-driver-definitions/src/tokenFetch.ts:
```diff
@@ -83,6 +99,29 @@ export const tokenFromResponse = (
+ * @returns Authorization header value
+ * @internal
+ */
+export const authHeaderFromResponse = (
```
authHeaderFromTokenResponse?

aboktor replied:
done

AndreiZe commented on packages/drivers/odsp-driver/src/fetchSnapshot.ts:
```diff
@@ -240,18 +232,16 @@ async function redeemSharingLink(
-						forceAccessTokenViaAuthorizationHeader,
+					const url = redeemUrl;
+					const method = "GET";
+					const storageToken = await getAuthHeader(
```
authHeader instead of storageToken

aboktor replied:
done

jatgarg commented on packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:
```diff
@@ -45,7 +47,7 @@ import {
  * is also referred to as the "VROOM" token in SPO.
  * @param persistedCache - Cache to store the fetched snapshot.
- * @param forceAccessTokenViaAuthorizationHeader - whether to force passing given token via authorization header.
+ * @param forceAccessTokenViaAuthorizationHeader - DEPRECATED (true value always used instead): whether to force passing given token via authorization header.
```
Please use @deprecated.

jatgarg commented on packages/drivers/odsp-driver-definitions/src/factory.ts:
```diff
@@ -64,6 +64,7 @@ export interface IOpsCachingPolicy {
  */
 export interface ICollabSessionOptions {
 	/**
+	 * @deprecated No longer needed
```
I think you wanted to deprecate the other option: forceAccessTokenViaAuthorizationHeader and not this one?

aboktor replied:
I want to deprecate both :)

AndreiZe reviewed (approved):
LGTM

aboktor committed:
Using @deprecated in jsdoc
Minor renaming
Other small changes addressing PR comments

pragya91 commented:
@aboktor Let's commit this change in main first, before porting it back to RC5, RC4, RC3.
FYI @jatgarg

pragya91 reviewed (dismissed):
@aboktor Let's commit this change in main first, before porting it back to RC5, RC4, RC3

aboktor committed:
Fixing mocha tests

mocka api update

pragya91 reviewed (commented):
.

aboktor commented:
> @aboktor Let's commit this change in main first, before porting it back to RC5, RC4, RC3

We talked 1:1 and agreed to go in RC3 first then port forward.

aboktor committed:
mocha-test-setup.api.md keeps changing, updating it one more time

vladsud commented on packages/drivers/odsp-driver-definitions/src/factory.ts:
```diff
@@ -64,6 +64,7 @@ export interface IOpsCachingPolicy {
  */
 export interface ICollabSessionOptions {
 	/**
+	 * @deprecated No longer needed
```
Could you please put (in comment, human readable - like RC3) a version when it's deprecated - simplifies the work to delete it in the future.

vladsud commented on packages/drivers/odsp-driver-definitions/src/tokenFetch.ts:
```diff
@@ -83,6 +99,29 @@ export const tokenFromResponse = (
+	if (
+		tokenResponse !== null &&
+		typeof tokenResponse === "object" &&
+		tokenResponse.authorizationHeader !== undefined
```
You can simplify it - tokenResponse?.authorizationHeader is sufficient. Everything is an object in JS

aboktor replied:
I can't do that because tokenResponse could be a string. But I can get rid of the null check because the typeof check will cover it.

vladsud commented on packages/drivers/odsp-driver/src/getFileLink.ts:
```diff
@@ -112,30 +112,35 @@ async function getFileLinkCore(
 				);
+				assert(
+					authHeader !== null,
+					0x2bb /* "Instrumented token fetcher with throwOnNullToken = true should never return null" */,
```
I do not think this assert does anything useful - getHeadersWithAuth() has same assert inside of it

vladsud commented on packages/drivers/odsp-driver/src/getUrlAndHeadersWithAuth.ts:
```diff
@@ -3,40 +3,18 @@
+	authHeader: string | null,
+): { [index: string]: string } {
+	assert(authHeader !== null && authHeader.length > 0, 0x936 /* should be token */);
+	if (!authHeader || authHeader.length === 0) {
```
You should choose one or the other - it does not make sense first to assert that authHeader is non-empty string, and then to have IF statement checking the same.
Also !authHeader checks for empty strings

aboktor commented on packages/drivers/odsp-driver/src/odspUtils.ts:
```diff
@@ -398,18 +405,18 @@ export function toInstrumentedOdspTokenFetcher(
 							});
 						}
-						if (token === null && throwOnNullToken) {
+						if (authHeader === null && throwOnNullToken) {
```
I will build the loop app and test against ODC with those changes. I don't expect problems here since I didn't change where tokens come from.

aboktor replied:
Tested Load and collab on ODC and it's working as expected.

aboktor committed:
Addressing comments from PR
- Adding comment about deprecation of unauthenticatedUserDisplayName
- Removed unnecessary assert
- Removed unnecessary checks after assert
- Simplified condition
